rpcserver.go

```diff
@@ -2544,226 +2574,478 @@ func marshalOutputType(outputType tappsbt.VOutputType) (taprpc.OutputType,
 	}
+
+	return handleEvents[bool, *taprpc.ReceiveEvent](
+		r.cfg.AssetCustodian, ntfnStream, marshaler, filter, r.quit,
+	)
 }
 
-// SubscribeReceiveAssetEventNtfns registers a subscription to the event
-// notification stream which relates to the asset receiving process.
-func (r *rpcServer) SubscribeReceiveAssetEventNtfns(
-	_ *taprpc.SubscribeReceiveAssetEventNtfnsRequest,
-	ntfnStream subReceiveEventNtfnsStream) error {
+// SubscribeSendEvents registers a subscription to the event notification
+// stream which relates to the asset sending process.
+func (r *rpcServer) SubscribeSendEvents(req *taprpc.SubscribeSendEventsRequest,
+	ntfnStream sendEventStream) error {
 
-	// Create a new event subscriber and pass a copy to the component(s)
-	// which are involved in the asset receive process. We will then read
-	// events from the subscriber.
+	var (
+		targetScriptKey *btcec.PublicKey
+		err             error
+	)
+	if len(req.FilterScriptKey) > 0 {
+		targetScriptKey, err = btcec.ParsePubKey(req.FilterScriptKey)
+		if err != nil {
+			return fmt.Errorf("error parsing script key: %w", err)
+		}
+	}
+
+	filter := func(event fn.Event) (bool, error) {
+		e, ok := event.(*tapfreighter.AssetSendEvent)
+		if !ok {
+			return false, fmt.Errorf("invalid event type: %T",
+				event)
+		}
+
+		// If we're not filtering on a specific script key, we return
+		// all events.
+		if targetScriptKey == nil {
+			return true, nil
+		}
+
+		// We can only match the script key on the active virtual
+		// packets, so if there are none, then this isn't an event we're
+		// interested in.
+		if len(e.VirtualPackets) == 0 {
+			return false, nil
+		}
+
+		for _, vPacket := range e.VirtualPackets {
+			for _, vOut := range vPacket.Outputs {
+				if vOut.ScriptKey.PubKey == nil {
+					continue
+				}
+
+				// This packet sends to the filtered script key,
+				// we want to include this event.
+				if vOut.ScriptKey.PubKey.IsEqual(
+					targetScriptKey,
+				) {
+
+					return true, nil
+				}
+			}
+		}
+
+		// No packets with outputs going to the filtered script key
+		// found.
+		return false, nil
+	}
+
+	return handleEvents[bool, *taprpc.SendEvent](
+		r.cfg.ChainPorter, ntfnStream, marshalSendEvent, filter, r.quit,
+	)
+}
+
+// handleEvents is a helper function that reads events from an event source and
+// forwards them to an RPC stream.
+func handleEvents[T any, Q any](eventSource fn.EventPublisher[fn.Event, T],
+	stream EventStream[Q], marshaler func(fn.Event) (Q, error),
+	filter func(fn.Event) (bool, error), quit <-chan struct{}) error {
+
+	// Create a new event subscriber and pass a copy to the event source.
+	// We will then read events from the subscriber.
 	eventSubscriber := fn.NewEventReceiver[fn.Event](fn.DefaultQueueSize)
 	defer eventSubscriber.Stop()
 
-	// Register the subscriber with the AssetCustodian.
-	err := r.cfg.AssetCustodian.RegisterSubscriber(
-		eventSubscriber, false, false,
-	)
+	var unused T
```

Comment:
nit: all small comment to explain why this isn't `nil` could be useful, took me a sec to figure it out.

Reply:
Makes sense, added.